src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    /// <summary>
+    /// Caches tokens on a per-line basis for documents based on semantic version.
+    /// </summary>
+    /// <remarks>
+    /// The cache makes the assumption that the tokens passed in are complete for the given range.
+    /// </remarks>
+    internal sealed class SemanticTokensCache
+    {
+        private const int MaxSemanticVersionPerDoc = 4;
+
+        // Nested cache mapping (URI -> (semanticVersion -> (line #s -> tokens on line)))
+        private readonly MemoryCache<DocumentUri, MemoryCache<VersionStamp, Dictionary<int, List<int>>>> _cache = new();
+
+        /// <summary>
```

Comment:
Like that you've included the full doc-comments for some of these methods that could feel a bit esoteric.